zebra-chain/src/orchard/keys.rs

```diff
@@ -0,0 +1,1008 @@
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct IvkCommitRandomness(pallas::Scalar);
+
+impl fmt::Debug for IvkCommitRandomness {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("IvkCommitRandomness")
+            .field(&hex::encode(self.0.to_bytes()))
+            .finish()
+    }
+}
+
+impl From<SpendingKey> for IvkCommitRandomness {
+    /// rivk = ToScalar^Orchard(PRF^expand_sk ([8]))
+    ///
+    /// https://zips.z.cash/protocol/protocol.pdf#orchardkeycomponents
+    fn from(sk: SpendingKey) -> Self {
+        let scalar = pallas::Scalar::from_bytes_wide(&prf_expand(sk.into(), vec![&[8]]));
+
+        Self(scalar)
+    }
+}
+
+impl From<IvkCommitRandomness> for [u8; 32] {
+    fn from(rivk: IvkCommitRandomness) -> Self {
+        rivk.0.into()
+    }
+}
+
+impl From<IvkCommitRandomness> for pallas::Scalar {
+    fn from(rivk: IvkCommitRandomness) -> Self {
+        rivk.0
+    }
+}
+
+impl TryFrom<[u8; 32]> for IvkCommitRandomness {
+    type Error = &'static str;
+
+    fn try_from(bytes: [u8; 32]) -> Result<Self, Self::Error> {
+        let possible_scalar = pallas::Scalar::from_bytes(&bytes);
+
+        if possible_scalar.is_some().into() {
+            Ok(Self(possible_scalar.unwrap()))
+        } else {
+            Err("Invalid pallas::Scalar value")
+        }
+    }
+}
+
+/// Magic human-readable strings used to identify what networks Orchard incoming
+/// viewing keys are associated with when encoded/decoded with bech32.
+///
+/// https://zips.z.cash/protocol/nu5.pdf#orchardinviewingkeyencoding
+mod ivk_hrp {
+    pub const MAINNET: &str = "zivko";
+    pub const TESTNET: &str = "zivktestorchard";
+}
+
+/// An incoming viewing key, as described in [protocol specification
+/// §4.2.3][ps].
+///
+/// Used to decrypt incoming notes without spending them.
+///
+/// [ps]: https://zips.z.cash/protocol/protocol.pdf#orchardkeycomponents
+#[derive(Copy, Clone, Eq, PartialEq)]
+pub struct IncomingViewingKey {
+    network: Network,
+    scalar: pallas::Scalar,
+}
+
+// TODO: impl a From that accepts a Network?
+
+impl fmt::Debug for IncomingViewingKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_tuple("IncomingViewingKey")
+            .field(&hex::encode(self.scalar.to_bytes()))
+            .finish()
+    }
+}
+
+impl fmt::Display for IncomingViewingKey {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        let hrp = match self.network {
+            Network::Mainnet => ivk_hrp::MAINNET,
+            _ => ivk_hrp::TESTNET,
+        };
+
+        bech32::encode_to_fmt(f, hrp, &self.scalar.to_bytes().to_base32(), Variant::Bech32).unwrap()
+    }
+}
+
+impl From<[u8; 32]> for IncomingViewingKey {
+    /// Generate an _IncomingViewingKey_ from existing bytes.
+    fn from(bytes: [u8; 32]) -> Self {
+        Self {
+            // TODO: handle setting the Network better.
+            network: Network::default(),
```

Comment:
Rename to `IncomingViewingKey::from_bytes` and add a `Network` argument.

Reply:
8b0094c8